### ISSUE TITLE
How do I open a new or existing file for random rw access without truncating it?

### ISSUE THREAD

**Issue by didactic-drunk** [crystal-lang/crystal#7849]
Mode:
- "r+" won't create new files
- "w" truncates
- "wb" truncates
- "w+" truncates
- "a+" no random writes

**Comment by oprypin** [comment]
> "a+" no random writes

I really doubt that :o
Might just need to reset to the start of the file.

**Comment by oprypin** [comment]
The issue tracker is not a good place for questions. Please check out some other options at https://crystal-lang.org/community/

**Comment by didactic-drunk** [comment]
> > "a+" no random writes
> 
> I really doubt that :o
> Might just need to reset to the start of the file.

It's a function of Posix when a file is opened with O_APPEND.  From the linux man page:
`Before each write(2), the file offset is positioned at the end of the file, as if with lseek(2).  The modification of the file offset and the write operation are performed as a single atomic step.`

Test code:
```crystal
File.open("test.txt", "a+") do |io|
  3.times do |i|
    io.seek 0
    io.puts "append #{i}"
  end
end
```

Expected output:
```
append 2
```

Actual output:
```
append 0
append 1
append 2
```

Run it again and the file length doubles rather than overwriting.

**Comment by didactic-drunk** [comment]
> The issue tracker is not a good place for questions. Please check out some other options at https://crystal-lang.org/community/

I'm not really asking how to do something with crystal as it stands but how to adjust the API to support common file operations.  Consider this a feature request or API bug report.

**Comment by ysbaddaden** [comment]
You must manually check whether the file exists or not, then choose to open with `w` (create file) or `r+` (read/write no truncate). There is no abstraction for what you want to achieve.

Please use the forum to ask questions. The diacussion may eventually lead to open an issue if there isn't a solution to the question.

**Comment by didactic-drunk** [comment]
Checking for a file first creates a race condition.  I need O_EXCL semantics.  There isn't a solution.

Maybe I wasn't specific enough.  My intention was to open a discussion on the File API by showing a security related deficiency in the current API and avoiding what happened when I submitted #7822 which while simple and seemingly necessary (how does Process.chroot function if you can't drop privileges?  [It doesn't](https://www.google.com/search?q=escaping+chroot+jail)) is sitting in limbo for another PR #7829 with no code to show for it while API design is discussed.  As #7829 shows there were numerous PR's submitting code to solve the process problem that mostly stagnated until they were closed.

So instead of submitting a PR allowing O_EXCL that may or may not be accepted, how do I start the discussion on what will be accepted as in PR#7829?

**Comment by jhass** [comment]
You identified a problem, good! Don't phrase the issue title as a question then, "Missing API to...". If you know a good lower level solution, describe it, in medium detail :) If you want to discuss a new high level API, actually propose  one :) "Let's add a new mode, `"a-"`", "Let's allow seeking back for `"a+"`", ...

**Comment by RX14** [comment]
Should this issue be closed in favour of #7857 then?
